base_location/models/res_city_zip.py

```diff
@@ -0,0 +1,47 @@
+# Copyright 2016 Nicolas Bessi, Camptocamp SA
+# Copyright 2018 Aitor Bouzas <aitor.bouzas@adaptivecity.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class BetterZip(models.Model):
+    """City/locations completion object"""
+
+    _name = "res.city.zip"
+    _description = __doc__
+    _order = "name asc"
+
+    name = fields.Char('ZIP')
+    city_id = fields.Many2one(
+        'res.city',
+        'City',
+        required=True,
+    )
+    latitude = fields.Float()
+    longitude = fields.Float()
+
+    @api.multi
+    @api.depends('name', 'city', 'state_id', 'country_id')
+    def name_get(self):
+        result = []
+        for rec in self:
+            name = []
+            if rec.name:
+                name.append('%(name)s' % {'name': rec.name})
+            name.append('%(name)s' % {'name': rec.city_id.name})
+            if rec.city_id.state_id:
+                name.append('%(name)s' % {'name': rec.city_id.state_id.name})
+            if rec.city_id.country_id:
+                name.append('%(name)s' % {'name': rec.city_id.country_id.name})
+            result.append((rec.id, ", ".join(name)))
+        return result
+
+    @api.model
+    def name_search(self, name='', args=None, operator='ilike', limit=100):
```

Comment:
I think changing `_rec_name` to `display_name` will perform the search over the new computed field, handling all these cases, and making it more performant.